In [31]:
#Import Libraries
import pandas as pd
import numpy as np
import os

In [32]:
#Path
path = r'C:\Users\Pink PC\Downloads\Instacart Basket Analysis'

In [33]:
#Import customers.csv
df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

In [34]:
df_customers.shape

(206209, 10)

In [35]:
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [36]:
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [37]:
df_customers.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [38]:
df_customers.head(100)

user_id  First Name    Surnam  Gender       STATE  Age date_joined  \
0     26711     Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1     33890    Patricia      Hart  Female  New Mexico   36    1/1/2017   
2     65803     Kenneth    Farley    Male       Idaho   35    1/1/2017   
3    125935    Michelle     Hicks  Female        Iowa   40    1/1/2017   
4    130797         Ann   Gilmore  Female    Maryland   26    1/1/2017   
..      ...         ...       ...     ...         ...  ...         ...   
95    86389     Stephen   Murillo    Male     Vermont   69    1/1/2017   
96   102956       Kathy  Mitchell  Female    Missouri   78    1/1/2017   
97   180919  Jacqueline     Roach  Female    Nebraska   21    1/1/2017   
98    74155       Karen      Berg  Female     Alabama   33    1/1/2017   
99    96166         NaN   Oconnor    Male    Oklahoma   51    1/1/2017   

    n_dependants                        fam_status  income  
0              3                           married  165665  
1              0                            single   59285  
2              2                           married   99568  
3              0                            single   42049  
4              1                           married   40374  
..           ...                               ...     ...  
95             3                           married   66800  
96             1                           married   96314  
97             3  living with parents and siblings   56831  
98             3                           married   95285  
99             1                           married  155673  

[100 rows x 10 columns]

In [39]:
#Drop the date_joined column
df_customers = df_customers.drop(columns = ['date_joined'])

In [40]:
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age  n_dependants  \
0    26711    Deborah  Esquivel  Female    Missouri   48             3   
1    33890   Patricia      Hart  Female  New Mexico   36             0   
2    65803    Kenneth    Farley    Male       Idaho   35             2   
3   125935   Michelle     Hicks  Female        Iowa   40             0   
4   130797        Ann   Gilmore  Female    Maryland   26             1   

  fam_status  income  
0    married  165665  
1     single   59285  
2    married   99568  
3     single   42049  
4    married   40374

In [41]:
#Rename columns for readability and intuitiveness
df_customers.rename(columns = {'Surnam' : 'Last Name'}, inplace = True)

In [42]:
df_customers.rename(columns = {'STATE' : 'State'}, inplace = True)

In [43]:
df_customers.rename(columns = {'n_dependants' : 'Number_of_Dependants'}, inplace = True)

In [44]:
df_customers.rename(columns = {'fam_status' : 'Familial_Status'}, inplace = True)

In [45]:
df_customers.head(100)

user_id  First Name Last Name  Gender       State  Age  \
0     26711     Deborah  Esquivel  Female    Missouri   48   
1     33890    Patricia      Hart  Female  New Mexico   36   
2     65803     Kenneth    Farley    Male       Idaho   35   
3    125935    Michelle     Hicks  Female        Iowa   40   
4    130797         Ann   Gilmore  Female    Maryland   26   
..      ...         ...       ...     ...         ...  ...   
95    86389     Stephen   Murillo    Male     Vermont   69   
96   102956       Kathy  Mitchell  Female    Missouri   78   
97   180919  Jacqueline     Roach  Female    Nebraska   21   
98    74155       Karen      Berg  Female     Alabama   33   
99    96166         NaN   Oconnor    Male    Oklahoma   51   

    Number_of_Dependants                   Familial_Status  income  
0                      3                           married  165665  
1                      0                            single   59285  
2                      2                           married   99568  
3                      0                            single   42049  
4                      1                           married   40374  
..                   ...                               ...     ...  
95                     3                           married   66800  
96                     1                           married   96314  
97                     3  living with parents and siblings   56831  
98                     3                           married   95285  
99                     1                           married  155673  

[100 rows x 9 columns]

In [46]:
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 9 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   user_id               206209 non-null  int64 
 1   First Name            194950 non-null  object
 2   Last Name             206209 non-null  object
 3   Gender                206209 non-null  object
 4   State                 206209 non-null  object
 5   Age                   206209 non-null  int64 
 6   Number_of_Dependants  206209 non-null  int64 
 7   Familial_Status       206209 non-null  object
 8   income                206209 non-null  int64 
dtypes: int64(4), object(5)
memory usage: 14.2+ MB


In [47]:
df_customers['user_id'] = df_customers['user_id'].astype('str')

In [48]:
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 9 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   user_id               206209 non-null  object
 1   First Name            194950 non-null  object
 2   Last Name             206209 non-null  object
 3   Gender                206209 non-null  object
 4   State                 206209 non-null  object
 5   Age                   206209 non-null  int64 
 6   Number_of_Dependants  206209 non-null  int64 
 7   Familial_Status       206209 non-null  object
 8   income                206209 non-null  int64 
dtypes: int64(3), object(6)
memory usage: 14.2+ MB


In [49]:
df_customers.describe()

Age  Number_of_Dependants         income
count  206209.000000         206209.000000  206209.000000
mean       49.501646              1.499823   94632.852548
std        18.480962              1.118433   42473.786988
min        18.000000              0.000000   25903.000000
25%        33.000000              0.000000   59874.000000
50%        49.000000              1.000000   93547.000000
75%        66.000000              3.000000  124244.000000
max        81.000000              3.000000  593901.000000

In [50]:
# Check for mixed types in df_customers
for column in df_customers.columns:
    print(column,':',pd.api.types.infer_dtype(df_customers[column]))

user_id : string
First Name : string
Last Name : string
Gender : string
State : string
Age : integer
Number_of_Dependants : integer
Familial_Status : string
income : integer


##There are no mixed-type columns

In [51]:
# Check for missing values
df_customers.isnull().sum()

user_id                     0
First Name              11259
Last Name                   0
Gender                      0
State                       0
Age                         0
Number_of_Dependants        0
Familial_Status             0
income                      0
dtype: int64

##There are 11,259 missing values in the First Name column.

In [52]:
#Create a new dataframe to view missing values
df_nan = df_customers[df_customers['First Name'].isnull()==True]

In [53]:
df_nan

user_id First Name Last Name  Gender         State  Age  \
53       76659        NaN   Gilbert    Male      Colorado   26   
73       13738        NaN     Frost  Female     Louisiana   39   
82       89996        NaN    Dawson  Female        Oregon   52   
99       96166        NaN   Oconnor    Male      Oklahoma   51   
105      29778        NaN    Dawson  Female          Utah   63   
...        ...        ...       ...     ...           ...  ...   
206038  121317        NaN    Melton    Male  Pennsylvania   28   
206044  200799        NaN  Copeland  Female        Hawaii   52   
206090  167394        NaN     Frost  Female        Hawaii   61   
206162  187532        NaN     Floyd  Female    California   39   
206171  116898        NaN   Delgado  Female      Colorado   23   

        Number_of_Dependants Familial_Status  income  
53                         2         married   41709  
73                         0          single   82518  
82                         3         married  117099  
99                         1         married  155673  
105                        3         married  151819  
...                      ...             ...     ...  
206038                     3         married   87783  
206044                     2         married  108488  
206090                     1         married   45275  
206162                     0          single   56325  
206171                     2         married   59222  

[11259 rows x 9 columns]

In [54]:
df_customers['First Name'].value_counts (dropna = False)

First Name
NaN        11259
Marilyn     2213
Barbara     2154
Todd        2113
Jeremy      2104
           ...  
Merry        197
Eugene       197
Garry        191
Ned          186
David        186
Name: count, Length: 208, dtype: int64

In [55]:
#Addressing missing values
df_customers['First Name'].fillna('Unknown', inplace = True)

In [56]:
#rechecking missing values
df_customers.isnull().sum()

user_id                 0
First Name              0
Last Name               0
Gender                  0
State                   0
Age                     0
Number_of_Dependants    0
Familial_Status         0
income                  0
dtype: int64

In [57]:
#Look for duplicate records
df_dups = df_customers[df_customers.duplicated()]

In [58]:
df_dups

Empty DataFrame
Columns: [user_id, First Name, Last Name, Gender, State, Age, Number_of_Dependants, Familial_Status, income]
Index: []

##There are no duplicate records in the dataframe. No action required. 

In [59]:
df_customers.shape

(206209, 9)

##Combine df_customers with orders_products_merged_updated_2.pkl 

In [60]:
#Import orders_products_merged_updated_2.pkl
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_derived.pkl'))

In [61]:
df_ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 26 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   product_id              int64   
 1   product_name            object  
 2   aisle_id                int64   
 3   department_id           int64   
 4   prices                  float64 
 5   order_id                int64   
 6   user_id                 int64   
 7   eval_set                object  
 8   order_number            int64   
 9   orders_day_of_week      int64   
 10  order_hour_of_day       int64   
 11  days_since_prior_order  float64 
 12  first_order             bool    
 13  add_to_cart_order       int64   
 14  reordered               int64   
 15  _merge                  category
 16  price_range_loc         object  
 17  busiest_day             object  
 18  busiest_days            object  
 19  busiest_period_of_day   object  
 20  max_order               int64   
 21  loyalt

In [62]:
#Change the data type of user_id to string
df_ords_prods_merge['user_id'] = df_ords_prods_merge['user_id'].astype('str')

In [63]:
#Change the data type of order_id to string
df_ords_prods_merge['order_id'] = df_ords_prods_merge['order_id'].astype('str')

In [64]:
df_ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 26 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   product_id              int64   
 1   product_name            object  
 2   aisle_id                int64   
 3   department_id           int64   
 4   prices                  float64 
 5   order_id                object  
 6   user_id                 object  
 7   eval_set                object  
 8   order_number            int64   
 9   orders_day_of_week      int64   
 10  order_hour_of_day       int64   
 11  days_since_prior_order  float64 
 12  first_order             bool    
 13  add_to_cart_order       int64   
 14  reordered               int64   
 15  _merge                  category
 16  price_range_loc         object  
 17  busiest_day             object  
 18  busiest_days            object  
 19  busiest_period_of_day   object  
 20  max_order               int64   
 21  loyalt

In [65]:
#Drop the _merge column from the df_ords_prods_merge dataframe as it is not needed. 
df_ords_prods_merge = df_ords_prods_merge.drop(columns = ['_merge'])

In [66]:
df_ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 25 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   product_id              int64  
 1   product_name            object 
 2   aisle_id                int64  
 3   department_id           int64  
 4   prices                  float64
 5   order_id                object 
 6   user_id                 object 
 7   eval_set                object 
 8   order_number            int64  
 9   orders_day_of_week      int64  
 10  order_hour_of_day       int64  
 11  days_since_prior_order  float64
 12  first_order             bool   
 13  add_to_cart_order       int64  
 14  reordered               int64  
 15  price_range_loc         object 
 16  busiest_day             object 
 17  busiest_days            object 
 18  busiest_period_of_day   object 
 19  max_order               int64  
 20  loyalty_flag            object 
 21  average spent           float

In [67]:
#Merge df_ords_prods_merge with df_customers
df_ords_prods_all = df_ords_prods_merge.merge(df_customers, on=['user_id'], indicator=True)

In [68]:
#Check the frequency of the new dataframe
df_ords_prods_all['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

In [69]:
df_ords_prods_all.head(20)

product_id                    product_name  aisle_id  department_id  \
0            1      Chocolate Sandwich Cookies        61             19   
1            1      Chocolate Sandwich Cookies        61             19   
2          907            Premium Sliced Bacon       106             12   
3          907            Premium Sliced Bacon       106             12   
4         1000                        Apricots        18             10   
5         3265   Mini Seedless Watermelon Pack        32              4   
6         4913            Table Water Crackers        78             19   
7         6184                     Clementines        32              4   
8         6901                     Raisin Bran       121             14   
9         7147  Classic Baby Creamers Potatoes        83              4   
10        7987      Hazelnut Spread with Cocoa        88             13   
11        8048         Packaged Grape Tomatoes        32              4   
12        8048         Packaged Grape Tomatoes        32              4   
13       10473             Half And Half Cream        53             16   
14       10473             Half And Half Cream        53             16   
15       10473             Half And Half Cream        53             16   
16       10473             Half And Half Cream        53             16   
17       10473             Half And Half Cream        53             16   
18       10473             Half And Half Cream        53             16   
19       10473             Half And Half Cream        53             16   

    prices order_id user_id eval_set  order_number  orders_day_of_week  ...  \
0      5.8  3139998     138    prior            28                   6  ...   
1      5.8  1977647     138    prior            30                   6  ...   
2     20.0  3160996     138    prior             1                   5  ...   
3     20.0  2254091     138    prior            10                   5  ...   
4     12.9   505689     138    prior             9                   6  ...   
5      1.7   960220     138    prior            17                   5  ...   
6      4.4   894221     138    prior            23                   5  ...   
7      4.3  1977647     138    prior            30                   6  ...   
8      2.9  1154289     138    prior            18                   1  ...   
9      9.0   930730     138    prior             2                   0  ...   
10    12.8  1592314     138    prior             5                   0  ...   
11    11.4   930730     138    prior             2                   0  ...   
12    11.4  1421595     138    prior             3                   2  ...   
13     7.0  3160996     138    prior             1                   5  ...   
14     7.0   930730     138    prior             2                   0  ...   
15     7.0  1421595     138    prior             3                   2  ...   
16     7.0  1592314     138    prior             5                   0  ...   
17     7.0  3052988     138    prior             6                   0  ...   
18     7.0  1986630     138    prior             7                   0  ...   
19     7.0  2823056     138    prior             8                   6  ...   

    order frequency flag  First Name  Last Name  Gender      State Age  \
0      Frequent customer     Charles        Cox    Male  Minnesota  81   
1      Frequent customer     Charles        Cox    Male  Minnesota  81   
2      Frequent customer     Charles        Cox    Male  Minnesota  81   
3      Frequent customer     Charles        Cox    Male  Minnesota  81   
4      Frequent customer     Charles        Cox    Male  Minnesota  81   
5      Frequent customer     Charles        Cox    Male  Minnesota  81   
6      Frequent customer     Charles        Cox    Male  Minnesota  81   
7      Frequent customer     Charles        Cox    Male  Minnesota  81   
8      Frequent customer     Charles        Cox    Male  Minnesota  81   
9      Frequent customer    

In [70]:
#8. Export this new dataframe as a pickle file.
df_ords_prods_all.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_all.pkl'))